### DISCUSSION TITLE
9월 19일 인터뷰 피드백

### DISCUSSION THREAD

**Discussion by devHudi** [woowacourse-study/2022-cs-plant#18] [General]
9월 19일 인터뷰 피드백을 남겨주세요.

**Comment by hyeonic** [comment]
## 리버 인터뷰

 * 1분 자기소개
 * 컨텍스트 스위칭 관련
 * 프로세스 간의 스레드 사이의 컨텍스트 스위칭은 의미 없는 것이 아닌가?

 * 멀티 프로세스는 동시성 문제가 일어나지 않을까?
 * 서블릿의 생명주기
 * 서블릿 왜쓸까?
 * 팬텀리드는 무엇일까?
 * 인터럽트?

 * 갑자기 톰캣 동시성 문제?

**Comment by devHudi** [comment]
# 리버

## **질문**

### 프로세스와 스레드 그리고 톰캣

[https://github.com/woowacourse-study/2022-cs-plant/issues/8](https://github.com/woowacourse-study/2022-cs-plant/issues/8)

- 스레드 컨텍스트 스위칭이 캐시 적중률이 높은 이유는 무엇인가요?
- 프로세스의 Code, Data 영역이 무엇인지 아시나요?
- 멀티 프로세스는 동시성 문제가 전혀 발생할 수 없을까요? 발생할 수 있다면 어떤 원인일까요?
    - 동시성 문제를 어떻게 방지할 수 있을까요?
        - 직접 이 문제를 해결한다면 어떻게 해결할까요?
    
    > IPC 통신, 커널 작업, System Call 등 (참고: [https://gyoogle.dev/blog/computer-science/operating-system/Race Condition.html](https://gyoogle.dev/blog/computer-science/operating-system/Race%20Condition.html))
    > 
- 인터럽트가 무엇인가요? 인터럽트가 발생하면 왜 컨텍스트 스위칭이 발생하나요?
- 컨텍스트 스위칭이 발생하면, 왜 CPU가 대기상태에 들어가나요?
    - 어차피 둘다 컴퓨터니까 처리 속도가 매우 빠를텐데 왜 CPU 혼자 쉬나요?
- 이벤트 드리븐이 무엇인가요?
- 톰캣이 성능 향상을 위해 멀티 프로세스 대신 멀티 쓰레드를 택한 이유는 무엇인가요?
    - Thread-Safe 가 무엇인가요?
    - 왜 예상치 못한 동작을 하게 될까요?
- 멀티 태스킹과 멀티 프로세싱의 차이를 아시나요?
    
    > 멀티 프로그래밍은 메모리에 여러 프로그램을 실행하는것. 멀티 프로세싱은 한번에 둘 이상의 프로세서(CPU)를 사용하는 것. 멀티 태스킹은 라운드 로빈등을 이용하여 한번에 여러작업을 실행하는것. 사실 컨텍스트 스위칭, 스케줄링 알고리즘 모두 멀티 프로세스 보다는 멀티 태스킹에 연관이 되어있음.
    > 
    
    > [https://www.geeksforgeeks.org/difference-between-multitasking-multithreading-and-multiprocessing/](https://www.geeksforgeeks.org/difference-between-multitasking-multithreading-and-multiprocessing/)
    > 

### 트랜잭션 격리 수준

[https://github.com/woowacourse-study/2022-cs-plant/issues/9](https://github.com/woowacourse-study/2022-cs-plant/issues/9)

- 트랜잭션을 왜 사용할까요?
- 반복 가능한 읽기에서 팬텀 리드가 발생하는 과정에 대해서 설명해주세요.
- 직렬화 가능 수준은 너무 느려서 쓸데 없어보이는데, 왜 사용할까요?
- MySQL에서는 반복 가능한 읽기가 팬텀 리드가 트랜잭션 별 스냅샷 덕분이라고 하셨는데, 이 과정을 자세히 설명해주세요.
    - 팬텀리드가 무엇인가요?
- 프로젝트에서 트랜잭션 격리 수준에 대해 고민해보신 적 있으신가요?
    - 프로젝트에서 트랜잭션 격리 수준을 적용한다면, 어떤 작업에 어떤 수준으로 적용하고 싶으신가요?

### 서블릿과 서블릿 생명주기

[https://github.com/woowacourse-study/2022-cs-plant/issues/17](https://github.com/woowacourse-study/2022-cs-plant/issues/17)

- 서블릿을 왜 사용할까요?
- 서블릿은 HTTP에서만 사용할 수 있나요?
    
    > HTTP에 국한되지 않고, 요청-응답 모델을 사용하는 모든 프로그램에서 사용할 수 있음
    > 
- 서블릿은 초기화 작업을 왜 생성자가 아닌 `init()` 에서 할까요?
    - (꼬리) 왜 생성자는 복잡한 비즈니스 로직을 가지면 안되는걸까요?
- 우테코 과정에서 우리는 톰캣을 간접적으로라도 사용해본 적 있을까요?
    
    > 스프링 부트의 내장 톰캣 서버에 대해 이야기하는 것을 의도
    > 
- 우테코에서 서블릿과 관련된 객체를 사용해보신 적 있나요?
    
    > Filter 등을 사용하면 `HttpServletRequest` 등을 사용해봤을수도 있음.
    > 
- 한번 요청을 처리한 서블릿은 쓰레드와 함께 소멸되죠?
    
    > 소멸 안됨
    > 
- 서블릿이 소멸되면 소멸된건데 왜 `destroy()` 라는 메소드가 있는 것일까요?

## **피드백**

- 프로세스 공유자원 경쟁상태 문제 해결방법에 대해 공부하지 않아도, 주어진 정보로 잘 추론해서 말해주었음.
- 흐름을 설명할때 생각할 시간이 조금 걸리더라도 생각을 조금 한뒤 말하면 좋을 것 같음.

**Comment by hyeonic** [comment]
## 서블릿 로딩 시점
 * https://www.inflearn.com/questions/405451
